core/src/main/scala/kafka/server/checkpoints/CheckpointFile.scala

```diff
@@ -58,11 +62,8 @@ class CheckpointFile[T](val file: File, version: Int, formatter: CheckpointFileF
         fileOutputStream.getFD().sync()
       } catch {
         case e: FileNotFoundException =>
-          if (FileSystems.getDefault.isReadOnly) {
-            fatal(s"Halting writes to checkpoint file (${file.getAbsolutePath}) because the underlying file system is inaccessible: ", e)
-            Exit.halt(1)
-          }
-          throw e
+          logDirFailureChannel.maybeAddLogFailureEvent(file.getParentFile.getParent)
+          throw new KafkaStorageException(s"Halting writes to checkpoint file (${file.getAbsolutePath}) because the underlying file system is inaccessible", e)
```

Comment:
I don't know why we don't halt the server on IOException when reading or writing the checkpoint file. Anyway,  I have updated the patch to catch IOException when reading/writing the checkpoint file, call maybeAddLogDirFailureEvent(), and re-throw a KafkaStorageException.